### PR TITLE
Ported contract address mining for Cow token

### PR DESCRIPTION
**Pull request by nlordell**
⚠️ The PR is marked as **DRAFT** because I don't know if we want to actually merge this or not. The code however is done as far as I can tell. I even added some unit tests to make sure the computed addresses match expectations.

---

This PR ports https://github.com/nlordell/deadbeef for use with the Cow token contracts.

Notably, it reads Safe and Cow token parameters from `examples/settings.json` and allows "mining" vanity addresses for the DAO Safe, team controller Safe and the COW Protocol token.

```
% cd vanity/
% # Make sure to build with `--release` for performance reasons.
% cargo build --release
...
    Finished release [optimized] target(s) in 15.30s
% ./target/release/vanity -s ../example/settings.json -p 0xca0da0 -c dao
salt nonce: 1725568918840712425654928017241366274873947264783149270648913780097139887465
% # COW DAO address affects final token address, so update settings with computed nonce:
% sed -i 's/"nonce": "123"/"nonce": "1725568918840712425654928017241366274873947264783149270648913780097139887465"/' ../example/settings.json
% ./target/release/vanity -s ../example/settings.json -p 0xca07ea -c team-controller
salt nonce: 114223000606167876931716684852058011638700684913499325987135106519965518060105
% ./target/release/vanity -s ../example/settings.json -p 0xca771e -c token
salt: 0x131b707afa9cfc44bed2ac14343051142853949f76dd917fd2defa8f57f23a79
```

### Test Plan

Using the computed nonces above:
```
% git diff
diff --git a/example/settings.json b/example/settings.json
index 482da3c..dcdf9b3 100644
--- a/example/settings.json
+++ b/example/settings.json
@@ -10,7 +10,7 @@
       "0x4444444444444444444444444444444444444444",
       "0x5555555555555555555555555555555555555555"
     ],
-    "nonce": "123"
+    "nonce": "1725568918840712425654928017241366274873947264783149270648913780097139887465"
   },
   "teamController": {
     "threshold": 2,
@@ -19,11 +19,11 @@
       "0x7777777777777777777777777777777777777777",
       "0x8888888888888888888888888888888888888888"
     ],
-    "nonce": "456"
+    "nonce": "114223000606167876931716684852058011638700684913499325987135106519965518060105"
   },
   "cowToken": {
     "expectedAddress": "0xcaddle0000000000000000000000000000000000000000000000000000001337",
-    "salt": "0x0000000000000000000000000000000000000000000000000000000000001337"
+    "salt": "0x131b707afa9cfc44bed2ac14343051142853949f76dd917fd2defa8f57f23a79"
   },
   "virtualCowToken": {
     "gnoPrice": "375000000000000",
```

Now compute a deployment:
```
% npx hardhat test-claims --user-count 10 --mnemonic "candy maple cake sugar pudding cream honey rich smooth crumble sweet treat"
...
% npx hardhat deployment --network mainnet --claims output/test-claims/claims.csv --settings example/settings.json
...
% cat output/deployment/addresses.json     
{
  "cowDao": "0xca0Da0d7C681b93B8BeEe34ef72C9b8Eb886A0cb",
  "teamController": "0xCA07eA4Db88C2d90B9D452d3E3716Bb021A418bF",
  "investorFundsTarget": "0xEd9897c84B4ed33FB66fF019136207E50eF6ACe0",
  "cowToken": "0xCA771eaB7fd9c3AB6a1618408834e62cADd01578"
}
```

Note how final contract addresses have the desired prefix!

### CODE REVIEW

**Comment by nlordell** [comment]
Also, just wanted to say the scripts all worked like a breeze (generating test claims and deployment) and were nicely composable. Great job y'all!

**Comment by nlordell** [comment]
2e85f9bd6496c05705bee5d84a5b34f567ef334f modifies the vanity address to use the `gnosisDao` field from the `settings.json` as the initial token holder.

File: vanity/src/safe.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    salt: [u8; 64],
+    create2: Create2,
+}
+
+impl Safe {
+    /// Creates a new safe from deployment parameters.
+    pub fn new(parameters: SafeParameters) -> Self {
+        // Safe requires sorted owners.
+        let mut owners = parameters.owners;
+        owners.sort_unstable();
```
Comment (fedgiac) [inline vanity/src/safe.rs:28]: I don't think this is true for creating a safe: https://github.com/gnosis/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L22-L43

It's true however for validating signature, it's done to make it cheaper to tell that there are no repeating addresses.

**Review by fedgiac — COMMENTED** [review]
Thanks for the great tool! It works like a charm. :smile:

File: vanity/src/safe.rs (hunk 1 above)
Comment (nlordell) [reply vanity/src/safe.rs:28]: Oops! 🙈

Comment (nlordell) [reply vanity/src/safe.rs]: Removed!

**Pull request closed without merging**